### PR TITLE
Add TK2 option to `FullPeepholeOptimise`

### PR DESCRIPTION
**Pull request by cqc-alec**

### CODE REVIEW

File: tket/src/Transformations/OptimisationPass.cpp [hunk 1]
```diff
@@ -37,11 +40,23 @@ Transform peephole_optimise_2q() {
...
+    case OpType::CX:
+      return (
+          synthesise_tket() >> two_qubit_squash() >>
+          clifford_simp(allow_swaps) >> synthesise_tket() >>
+          three_qubit_squash() >> clifford_simp(allow_swaps) >>
+          synthesise_tket());
+    case OpType::TK2:
+      return (
+          synthesise_tk() >> two_qubit_squash(OpType::TK2) >>
+          clifford_simp(allow_swaps) >> synthesise_tk() >>
```
Comment (lmondada) [inline tket/src/Transformations/OptimisationPass.cpp:54]: This should take care of Cliffords properly: any `TK2` gate that was clifford and decomposed will be resynthesised after `clifford_simp`. Swaps should be swallowed within TK2s and be handled later.

```suggestion
          clifford_simp(false) >> two_qubit_squash(OpType::TK2) >> synthesise_tk() >>
```

File: tket/tests/test_CompilerPass.cpp [hunk 2]
```diff
@@ -771,6 +772,57 @@ SCENARIO("PeepholeOptimise2Q and FullPeepholeOptimise") {
...
+    unsigned n_gates_noswaps_tk2 = compiled_circ_noswaps_tk2.n_gates();
+    unsigned n_tk2_noswaps_tk2 =
+        compiled_circ_noswaps_tk2.count_gates(OpType::TK2);
+    unsigned n_tk1_noswaps_tk2 =
+        compiled_circ_noswaps_tk2.count_gates(OpType::TK1);
+
+    CHECK(n_gates_swaps_cx == n_cx_swaps_cx + n_tk1_swaps_cx);
+    CHECK(n_gates_swaps_tk2 == n_tk2_swaps_tk2 + n_tk1_swaps_tk2);
+    CHECK(n_gates_noswaps_cx == n_cx_noswaps_cx + n_tk1_noswaps_cx);
+    CHECK(n_gates_noswaps_tk2 == n_tk2_noswaps_tk2 + n_tk1_noswaps_tk2);
```
[the remark this answers is not in the document]
Comment (lmondada) [reply tket/tests/test_CompilerPass.cpp:822]: Mmmm, yes that's true. I guess that in some circuits with specific structure, the gains in terms of `Rzz` gates might make the effort worth it.

Any chance that random circuits are particularly ill-suited?

**Review by lmondada — CHANGES_REQUESTED** [review]
All good. I suggest adding `two_qubit_squash` after `clifford_simp` as I mentioned. This should take care of TK2 support in `clifford_simp`, so we won't need a separate PR.

File: tket/tests/test_CompilerPass.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (cqc-alec) [reply tket/tests/test_CompilerPass.cpp:822]: > Any chance that random circuits are particularly ill-suited?

Yes, that is possible.

**Review by lmondada — APPROVED** [review]

**Pull request merged**